Events not always gossipped to all nodes

With a gossip fan out smaller than the cluster size, it's possible that some gossip updates are not propagated to the whole cluster if nodes keep picking the same nodes over and over.

During our test we get the following gossip pattern:

`zeebe-0`:
```
13:00:29.630 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 0 - Gossipping updates [Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846567, state=ALIVE, incarnationNumber=1561033846573}] to Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}}
13:00:29.634 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 0 - Gossipping updates [Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846567, state=ALIVE, incarnationNumber=1561033846573}] to Member{id=1, address=zeebe-1.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":1,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-1.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":true,"2":true,"5":false,"6":false,"7":true}}}}
```

`zeebe-1`:
```
13:00:29.751 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 1 - Gossipping updates [Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846567, state=ALIVE, incarnationNumber=1561033846573}] to Member{id=2, address=zeebe-2.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":2,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-2.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"2":false,"3":true,"6":true,"7":false,"8":false}}}}
13:00:29.751 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 1 - Gossipping updates [Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846567, state=ALIVE, incarnationNumber=1561033846573}] to Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}}
```

`zeebe-2`:
```
13:00:29.924 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 2 - Gossipping updates [Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846567, state=ALIVE, incarnationNumber=1561033846573}] to Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}}
13:00:29.927 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 2 - Gossipping updates [Member{id=4, address=zeebe-4.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":4,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-4.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"3":false,"4":false,"5":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846567, state=ALIVE, incarnationNumber=1561033846573}] to Member{id=1, address=zeebe-1.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":1,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-1.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":true,"2":true,"5":false,"6":false,"7":true}}}}
```

`zeebe-3`:
```
```

`zeebe-4`:
```
13:00:29.820 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 4 - Gossipping updates [Member{id=0, address=zeebe-0.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":0,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-0.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"4":true,"5":true,"6":false}}}, version=3.2.0-alpha1, timestamp=1561033847627, state=ALIVE, incarnationNumber=1561033847630}] to Member{id=0, address=zeebe-0.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":0,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-0.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"4":true,"5":true,"6":false}}}}
13:00:33.820 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 4 - Gossipping updates [Member{id=2, address=zeebe-2.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":2,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-2.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"2":false,"3":true,"6":true,"7":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846670, state=ALIVE, incarnationNumber=1561033846672}] to Member{id=2, address=zeebe-2.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":2,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-2.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"2":false,"3":true,"6":true,"7":false,"8":false}}}}
13:00:33.822 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 4 - Gossipping updates [Member{id=2, address=zeebe-2.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":2,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-2.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"2":false,"3":true,"6":true,"7":false,"8":false}}}, version=3.2.0-alpha1, timestamp=1561033846670, state=ALIVE, incarnationNumber=1561033846672}] to Member{id=0, address=zeebe-0.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":0,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-0.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"4":true,"5":true,"6":false}}}}
13:00:34.820 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 4 - Gossipping updates [Member{id=1, address=zeebe-1.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":1,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-1.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":true,"2":true,"5":false,"6":false,"7":true}}}, version=3.2.0-alpha1, timestamp=1561033846747, state=ALIVE, incarnationNumber=1561033846750}] to Member{id=0, address=zeebe-0.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":0,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-0.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":false,"4":true,"5":true,"6":false}}}}
13:00:34.821 [] [atomix-cluster-heartbeat-sender] TRACE io.atomix.cluster.protocol.SwimMembershipProtocol - 4 - Gossipping updates [Member{id=1, address=zeebe-1.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":1,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-1.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":true,"2":true,"5":false,"6":false,"7":true}}}, version=3.2.0-alpha1, timestamp=1561033846747, state=ALIVE, incarnationNumber=1561033846750}] to Member{id=1, address=zeebe-1.zeebe.npepinpe.svc.cluster.local:26502, properties={brokerInfo={"nodeId":1,"partitionsCount":8,"clusterSize":5,"replicationFactor":3,"addresses":{"command":"zeebe-1.zeebe.npepinpe.svc.cluster.local:26501"},"partitionRoles":{"1":true,"2":true,"5":false,"6":false,"7":true}}}}
```

I'd like to propose we work on fixing this again. I pushed a branch with a horrible test, but it highlights the problem - if you run it in a loop until failure, it will eventually fail as one of the nodes is missing some updates.

https://github.com/zeebe-io/atomix/commit/c57a216fd9b8d771997b5decde974237011b5d71

Basically, eventually you will see an error like:

```
java.lang.AssertionError: Unexpected event null

	at io.atomix.cluster.protocol.SwimProtocolTest.checkEvents(SwimProtocolTest.java:337)
	at io.atomix.cluster.protocol.SwimProtocolTest.shouldEventuallyPropagateUpdates(SwimProtocolTest.java:239)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

Which means we were expecting some event (e.g. member3 rejoining) but that node never received anything.

--- a/cluster/src/main/java/io/atomix/cluster/protocol/SwimMembershipProtocol.java
+++ b/cluster/src/main/java/io/atomix/cluster/protocol/SwimMembershipProtocol.java
@@ -15,6 +15,26 @@
  */
 package io.atomix.cluster.protocol;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static io.atomix.utils.concurrent.Threads.namedThreads;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import io.atomix.cluster.BootstrapService;
+import io.atomix.cluster.Member;
+import io.atomix.cluster.MemberId;
+import io.atomix.cluster.Node;
+import io.atomix.cluster.discovery.NodeDiscoveryEvent;
+import io.atomix.cluster.discovery.NodeDiscoveryEventListener;
+import io.atomix.cluster.discovery.NodeDiscoveryService;
+import io.atomix.cluster.impl.AddressSerializer;
+import io.atomix.utils.Version;
+import io.atomix.utils.event.AbstractListenerManager;
+import io.atomix.utils.net.Address;
+import io.atomix.utils.serializer.Namespace;
+import io.atomix.utils.serializer.Namespaces;
+import io.atomix.utils.serializer.Serializer;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -36,31 +56,10 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import java.util.stream.Collectors;
-
-import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Lists;
-import com.google.common.collect.Maps;
-import io.atomix.cluster.BootstrapService;
-import io.atomix.cluster.Member;
-import io.atomix.cluster.MemberId;
-import io.atomix.cluster.Node;
-import io.atomix.cluster.discovery.NodeDiscoveryEvent;
-import io.atomix.cluster.discovery.NodeDiscoveryEventListener;
-import io.atomix.cluster.discovery.NodeDiscoveryService;
-import io.atomix.cluster.impl.AddressSerializer;
-import io.atomix.utils.Version;
-import io.atomix.utils.event.AbstractListenerManager;
-import io.atomix.utils.net.Address;
-import io.atomix.utils.serializer.Namespace;
-import io.atomix.utils.serializer.Namespaces;
-import io.atomix.utils.serializer.Serializer;
 import org.apache.commons.lang3.tuple.ImmutablePair;
 import org.apache.commons.lang3.tuple.Pair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import static com.google.common.base.MoreObjects.toStringHelper;
-import static io.atomix.utils.concurrent.Threads.namedThreads;
 
 /**
  * SWIM group membership protocol implementation.
@@ -141,6 +140,7 @@ public class SwimMembershipProtocol
   private List<SwimMember> randomMembers = Lists.newCopyOnWriteArrayList();
   private final NodeDiscoveryEventListener discoveryEventListener = this::handleDiscoveryEvent;
   private final Map<MemberId, ImmutableMember> updates = new LinkedHashMap<>();
+  private final List<SwimMember> syncMembers = new ArrayList<>();
 
   private final ScheduledExecutorService swimScheduler = Executors.newSingleThreadScheduledExecutor(
       namedThreads("atomix-cluster-heartbeat-sender", LOGGER));
@@ -148,6 +148,7 @@ public class SwimMembershipProtocol
       namedThreads("atomix-cluster-events", LOGGER));
   private ScheduledFuture<?> gossipFuture;
   private ScheduledFuture<?> probeFuture;
+  private ScheduledFuture<?> syncFuture;
 
   private final AtomicInteger probeCounter = new AtomicInteger();
 
@@ -343,7 +344,7 @@ public class SwimMembershipProtocol
   /**
    * Synchronizes the node state with peers.
    */
-  private void sync() {
+  private void syncAll() {
     List<SwimMember> syncMembers = discoveryService.getNodes().stream()
         .map(node -> new SwimMember(MemberId.from(node.id().id()), node.address()))
         .filter(member -> !member.id().equals(localMember.id()))
@@ -359,17 +360,33 @@ public class SwimMembershipProtocol
    * @param member the peer with which to synchronize the node state
    */
   private void sync(ImmutableMember member) {
-    LOGGER.trace("{} - Synchronizing membership with {}", localMember.id(), member);
+    LOGGER.debug("{} - Synchronizing membership with {}", localMember.id(), member);
     bootstrapService.getMessagingService().sendAndReceive(
         member.address(), MEMBERSHIP_SYNC, SERIALIZER.encode(localMember.copy()), false, config.getProbeTimeout())
         .whenCompleteAsync((response, error) -> {
           if (error == null) {
             Collection<ImmutableMember> members = SERIALIZER.decode(response);
+            LOGGER.debug("{} - Synchronized membership with {}, received: {}", localMember.id(), member, members);
             members.forEach(this::updateState);
           } else {
             LOGGER.debug("{} - Failed to synchronize membership with {}", localMember.id(), member);
           }
         }, swimScheduler);
+  }
+
+  private void sync() {
+    if (syncMembers.isEmpty()) {
+      syncMembers.addAll(members.values());
+      syncMembers.remove(localMember);
+      Collections.shuffle(syncMembers);
+    }
+
+    if (!syncMembers.isEmpty()) {
+      final SwimMember member = syncMembers.remove(0);
+      if (member != null) {
+        sync(member.copy());
+      }
+    }
   }
 
   /**
@@ -730,7 +747,10 @@ public class SwimMembershipProtocol
           this::gossip, 0, config.getGossipInterval().toMillis(), TimeUnit.MILLISECONDS);
       probeFuture = swimScheduler.scheduleAtFixedRate(
           this::probe, 0, config.getProbeInterval().toMillis(), TimeUnit.MILLISECONDS);
-      swimScheduler.execute(this::sync);
+      swimScheduler.execute(this::syncAll);
+      syncFuture =
+          swimScheduler.scheduleAtFixedRate(
+              this::sync, 0, config.getSyncInterval().toMillis(), TimeUnit.MILLISECONDS);
       LOGGER.info("Started");
     }
     return CompletableFuture.completedFuture(null);
@@ -742,6 +762,7 @@ public class SwimMembershipProtocol
       discoveryService.removeListener(discoveryEventListener);
       gossipFuture.cancel(false);
       probeFuture.cancel(false);
+      syncFuture.cancel(false);
       swimScheduler.shutdownNow();
       eventExecutor.shutdownNow();
       LOGGER.info("{} - Member deactivated: {}", localMember.id(), localMember);

--- a/cluster/src/main/java/io/atomix/cluster/protocol/SwimMembershipProtocolConfig.java
+++ b/cluster/src/main/java/io/atomix/cluster/protocol/SwimMembershipProtocolConfig.java
@@ -15,10 +15,10 @@
  */
 package io.atomix.cluster.protocol;
 
-import java.time.Duration;
-
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.time.Duration;
 
 /**
  * SWIM membership protocol configuration.
@@ -33,6 +33,7 @@ public class SwimMembershipProtocolConfig extends GroupMembershipProtocolConfig 
   private static final int DEFAULT_PROBE_TIMEOUT = 2000;
   private static final int DEFAULT_SUSPECT_PROBES = 3;
   private static final int DEFAULT_FAILURE_TIMEOUT = 10000;
+  private static final int DEFAULT_SYNC_INTERVAL = 10_000;
 
   private boolean broadcastUpdates = DEFAULT_BROADCAST_UPDATES;
   private boolean broadcastDisputes = DEFAULT_BROADCAST_DISPUTES;
@@ -43,6 +44,7 @@ public class SwimMembershipProtocolConfig extends GroupMembershipProtocolConfig 
   private Duration probeTimeout = Duration.ofMillis(DEFAULT_PROBE_TIMEOUT);
   private int suspectProbes = DEFAULT_SUSPECT_PROBES;
   private Duration failureTimeout = Duration.ofMillis(DEFAULT_FAILURE_TIMEOUT);
+  private Duration syncInterval = Duration.ofMillis(DEFAULT_SYNC_INTERVAL);
 
   /**
    * Returns whether to broadcast member updates to all peers.
@@ -229,6 +231,29 @@ public class SwimMembershipProtocolConfig extends GroupMembershipProtocolConfig 
     checkNotNull(failureTimeout, "failureTimeout cannot be null");
     checkArgument(!failureTimeout.isNegative() && !failureTimeout.isZero(), "failureTimeout must be positive");
     this.failureTimeout = checkNotNull(failureTimeout);
+    return this;
+  }
+
+  /**
+   * Returns the sync interval.
+   *
+   * @return the sync interval
+   */
+  public Duration getSyncInterval() {
+    return syncInterval;
+  }
+
+  /**
+   * Sets the sync interval.
+   *
+   * @param syncInterval the sync interval
+   * @return the membership configuration
+   */
+  public SwimMembershipProtocolConfig setSyncInterval(Duration syncInterval) {
+    checkNotNull(syncInterval, "syncInterval cannot be null");
+    checkArgument(
+        !syncInterval.isNegative() && !syncInterval.isZero(), "syncInterval must be positive");
+    this.syncInterval = syncInterval;
     return this;
   }
 

--- a/cluster/src/test/java/io/atomix/cluster/protocol/SwimProtocolTest.java
+++ b/cluster/src/test/java/io/atomix/cluster/protocol/SwimProtocolTest.java
@@ -15,6 +15,12 @@
  */
 package io.atomix.cluster.protocol;
 
+import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.MEMBER_ADDED;
+import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.MEMBER_REMOVED;
+import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.METADATA_CHANGED;
+import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.REACHABILITY_CHANGED;
+import static org.junit.Assert.assertEquals;
+
 import com.google.common.collect.HashMultiset;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Multiset;
@@ -33,10 +39,6 @@ import io.atomix.cluster.messaging.impl.TestMessagingServiceFactory;
 import io.atomix.cluster.messaging.impl.TestUnicastServiceFactory;
 import io.atomix.utils.Version;
 import io.atomix.utils.net.Address;
-import net.jodah.concurrentunit.ConcurrentTestCase;
-import org.junit.Before;
-import org.junit.Test;
-
 import java.time.Duration;
 import java.util.Arrays;
 import java.util.Collection;
@@ -45,12 +47,10 @@ import java.util.Properties;
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.TimeUnit;
-
-import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.MEMBER_ADDED;
-import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.MEMBER_REMOVED;
-import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.METADATA_CHANGED;
-import static io.atomix.cluster.protocol.GroupMembershipEvent.Type.REACHABILITY_CHANGED;
-import static org.junit.Assert.assertEquals;
+import java.util.function.UnaryOperator;
+import net.jodah.concurrentunit.ConcurrentTestCase;
+import org.junit.Before;
+import org.junit.Test;
 
 /**
  * SWIM membership protocol test.
@@ -192,9 +192,75 @@ public class SwimProtocolTest extends ConcurrentTestCase {
     checkEvent(member2, MEMBER_ADDED, member);
   }
 
+  @Test
+  public void shouldSynchronizePeriodically() throws InterruptedException {
+    // given
+    final Duration gossipInterval = Duration.ofMillis(100);
+    final Duration syncInterval = Duration.ofMillis(500);
+    startProtocol(member1, c -> c.setSyncInterval(syncInterval).setGossipInterval(gossipInterval));
+    startProtocol(member2, c -> c.setSyncInterval(syncInterval).setGossipInterval(gossipInterval));
+    final SwimMembershipProtocol protocol3 =
+        startProtocol(
+            member3, c -> c.setSyncInterval(syncInterval).setGossipInterval(gossipInterval));
+
+    // wait for all nodes to know about each other
+    checkEvents(
+        member1,
+        new GroupMembershipEvent(MEMBER_ADDED, member1),
+        new GroupMembershipEvent(MEMBER_ADDED, member2),
+        new GroupMembershipEvent(MEMBER_ADDED, member3));
+    checkEvents(
+        member2,
+        new GroupMembershipEvent(MEMBER_ADDED, member1),
+        new GroupMembershipEvent(MEMBER_ADDED, member2),
+        new GroupMembershipEvent(MEMBER_ADDED, member3));
+    checkEvents(
+        member3,
+        new GroupMembershipEvent(MEMBER_ADDED, member1),
+        new GroupMembershipEvent(MEMBER_ADDED, member2),
+        new GroupMembershipEvent(MEMBER_ADDED, member3));
+
+    // when
+    // isolate member3
+    partition(member3);
+    checkEvents(member1, new GroupMembershipEvent(REACHABILITY_CHANGED, member3), new GroupMembershipEvent(MEMBER_REMOVED, member3));
+    checkEvents(member2, new GroupMembershipEvent(REACHABILITY_CHANGED, member3), new GroupMembershipEvent(MEMBER_REMOVED, member3));
+    checkEvents(
+        member3,
+        new GroupMembershipEvent(REACHABILITY_CHANGED, member1),
+        new GroupMembershipEvent(MEMBER_REMOVED, member1),
+        new GroupMembershipEvent(REACHABILITY_CHANGED, member2),
+        new GroupMembershipEvent(MEMBER_REMOVED, member2));
+
+    // update member1 and wait for the property to be propagated
+    member1.properties().put("newProperty", 1);
+    checkEvents(member1, new GroupMembershipEvent(METADATA_CHANGED, member1));
+    checkEvents(member2, new GroupMembershipEvent(METADATA_CHANGED, member1));
+
+    // ensure member2 has already tried to propagate the new property, then reconnect it to member3
+    // it shouldn't try to update it with member1, and member1 is disconnected from member3 so will
+    // not send it probe requests - the only way for member3 to receive the new property is for it
+    // to sync with member2
+    Thread.sleep(gossipInterval.toMillis());
+    heal(member2, member3);
+    checkEvent(member2, MEMBER_ADDED, member3);
+    checkEvents(member3, new GroupMembershipEvent(MEMBER_ADDED, member2));
+
+    // then
+    // wait until member3 has tried to sync
+    Thread.sleep(syncInterval.toMillis());
+    assertEquals(1, protocol3.getMember(member1.id()).properties().get("newProperty"));
+  }
+
   private SwimMembershipProtocol startProtocol(Member member) {
-    SwimMembershipProtocol protocol = new SwimMembershipProtocol(new SwimMembershipProtocolConfig()
-        .setFailureTimeout(Duration.ofSeconds(2)));
+    return startProtocol(member, UnaryOperator.identity());
+  }
+
+  private SwimMembershipProtocol startProtocol(Member member, UnaryOperator<SwimMembershipProtocolConfig> configurator) {
+    SwimMembershipProtocol protocol =
+        new SwimMembershipProtocol(
+            configurator.apply(
+                new SwimMembershipProtocolConfig().setFailureTimeout(Duration.ofSeconds(2))));
     TestGroupMembershipEventListener listener = new TestGroupMembershipEventListener();
     listeners.put(member.id(), listener);
     protocol.addListener(listener);
@@ -247,7 +313,7 @@ public class SwimProtocolTest extends ConcurrentTestCase {
     for (int i = 0; i < types.length; i++) {
       GroupMembershipEvent event = nextEvent(member);
       if (!events.remove(event)) {
-        throw new AssertionError();
+        throw new AssertionError("Unexpected event " + event);
       }
     }
   }

